### PR TITLE
cmd/kubeadm: use secure port for liveness probe on controller-manager and scheduler

### LINKED ISSUE #1327: Use secure port for scheduler and controller-manager liveness probe
## Feature request: 

Since 1.13, `kube-controller-manager` and `kube-scheduler` expose secure ports and marked insecure ports as deprecated.

scheduler: https://github.com/kubernetes/kubernetes/pull/69663

	change: 1.13
	enable secure port: 10259
	deprecate old port: 10251

controller-manager: https://github.com/kubernetes/kubernetes/pull/67069

	change: 1.12
	enable secure port: 10257
	deprecate old port: 10252

We should use the secure ports as the default the livenessProbes going forward. I have this fixed in a fork and will open a PR.

## Versions

**kubeadm version** 

```
kubeadm version: &version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.0-alpha.0.1048+21904a5f7d38b1", GitCommit:"21904a5f7d38b1054edad36f3b05b0d7bd556d08", GitTreeState:"clean", BuildDate:"2018-12-18T01:09:09Z", GoVersion:"go1.11.3", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by joshrosso**
/kind feature

**What this PR does / why we need it**:

This PR uses the secure port (tls) for liveness (/healthz) checks
against the kube-controller-manager and kube-scheduler. This makes the
livenessProbe protocol consistent between the kube-apiserver,
kube-controller-manager, and kube-scheduler.

Secure ports were introduced in 1.12 for the kube-controller-manager[0]
and 1.13 for the kube-scheduler[1]. The insecure ports have since been
deprecated.

[0]: https://github.com/kubernetes/kubernetes/pull/67069
[1]: https://github.com/kubernetes/kubernetes/pull/69663

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1327

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm defaults to secure ports of kube-controller-manager and kube-scheduler for livenessProbe.
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
thank you for this PR @joshrosso 
/lgtm
/priority important-longterm
/ok-to-test

@kubernetes/sig-cluster-lifecycle-pr-reviews 
/assign @timothysc @fabriziopandini 
for approve

**Comment by neolit123** [comment]
@joshrosso 
might be a good idea to include the port numbers in the user facing note (release note), so that users can quickly verify their setups.

File: cmd/kubeadm/app/preflight/checks.go [hunk 1]
```diff
@@ -878,6 +878,8 @@ func RunInitMasterChecks(execer utilsexec.Interface, cfg *kubeadmapi.InitConfigu
 		PortOpenCheck{port: int(cfg.LocalAPIEndpoint.BindPort)},
 		PortOpenCheck{port: 10251},
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/preflight/checks.go]: PortOpenCheck for 10251 and 10252 is still necessary?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ func GetStaticPodSpecs(cfg *kubeadmapi.InitConfiguration, k8sVersion *version.Ve
 			ImagePullPolicy: v1.PullIfNotPresent,
 			Command:         getControllerManagerCommand(cfg, k8sVersion),
 			VolumeMounts:    staticpodutil.VolumeMountMapToSlice(mounts.GetVolumeMounts(kubeadmconstants.KubeControllerManager)),
-			LivenessProbe:   staticpodutil.ComponentProbe(cfg, kubeadmconstants.KubeControllerManager, 10252, "/healthz", v1.URISchemeHTTP),
+			LivenessProbe:   staticpodutil.ComponentProbe(cfg, kubeadmconstants.KubeControllerManager, 10257, "/healthz", v1.URISchemeHTTPS),
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: From the documentation https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/ it seems that when using HTTPS probes with hostnetwork=true, it is necessary to set the host header.
Is this the case for kubeadm as well or not?

Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Please fix also the test cases for self-hosting pod mutation for using new port numbers (see `cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go`)

**Review by fabriziopandini — COMMENTED** [review]
@joshrosso thanks for your first-time contributions on kubeadm!
Only few minor comments from my side

**Review by luxas — COMMENTED** [review]
Hi, thanks for this contribution.
If I understand this correctly though, we actually need to configure kcm and the scheduler to start serving securely first, by setting `--authentication-kubeconfig` and `--authorization-kubeconfig` to the same value as `--kubeconfig`. Additionally, we need to generate serving certificates for both components so they serve a HTTPS endpoint using the same CA as the rest of the cluster. So this gets fairly more complicated than this :thinking: 

@sttts @liggitt am I right here?

**Comment by joshrosso** [comment]
That's interesting. My lab testing has shown that the `/healthz` endpoint is accessible through the secure port.

**example from the kcm**
```
curl https://10.30.0.12:10257/healthz -k

ok
```

I'll have more time later in the week to investigate this further and make changes if necessary.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (joshrosso) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @fabianofranz, yes i can.

It looks as if the selfhosting_test hasn't seen much work recently as it's still running `1.7.4`. So changes to the test cases will involve more work. To verify, are changes that large appropriate for this PR?

File: cmd/kubeadm/app/preflight/checks.go (hunk 1 above)
Comment (joshrosso) [reply cmd/kubeadm/app/preflight/checks.go]: The ports are still accessible (even when unused), at least in my 13.0 testing.

I'm unsure whether kcm and scheduler fail if these ports can't be opened. I'll test that later this week.

**Comment by joshrosso** [comment]
@luxas does this answer your question(s)?

flag on kcm & scheduler:
```

--tls-cert-file string

File  containing the default x509 Certificate for HTTPS. (CA cert, if any,  concatenated after server cert). 
If HTTPS serving is enabled, and  --tls-cert-file and --tls-private-key-file are not provided, a  self-signed 
certificate and key are generated for the public address and  saved to the directory specified by --cert-dir.
```

**Comment by liggitt** [comment]
`curl -k ...` and kubelet liveness checks tolerate self-signed TLS certs

`/healthz` is specifically allowed to be accessed even if delegated authn/authz isn't configured

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: I think that changing only the `bind-address` bit is ok
All the rest should go into another PR (or at least an issue should be opened to track this)

**Comment by joshrosso** [comment]
ref: https://github.com/kubernetes/kubeadm/issues/1285

I have time carved out this week to work through the comments.

**Comment by luxas** [comment]
> ref: kubernetes/kubeadm#1285

Yes, I mean we should enable the secure ports properly first (and test it properly!), before starting to rely on this feature.

**Comment by luxas** [comment]
a) make sure the secure serving cert is using the cluster CA
b) make sure that serving cert is rotated on upgrades
c) add a phases command for it
d) set --authentication-kubeconfig and --authorization-kubeconfig
e) make sure the correct RBAC rules are in place and used
f) disable the insecure port and _then_ switch over the liveness probes

File: cmd/kubeadm/app/preflight/checks.go (hunk 1 above)
Comment (timothysc) [inline cmd/kubeadm/app/preflight/checks.go]: @joshrosso - any update?

**Review by timothysc — APPROVED** [review]
/approve

File: cmd/kubeadm/app/preflight/checks.go (hunk 1 above)
Comment (joshrosso) [reply cmd/kubeadm/app/preflight/checks.go]: @timothysc not yet. I think this PR is blocked on the changes luxas suggested in an above comment.

I know what needs to be done, I just need to find the time to do them. I won’t have any availability for about 2 weeks. But would like to tackle this soon if no one gets it before me.

**Comment by alexbrand** [comment]
Not sure if this should be part of this PR, but another change that I think needs to happen is to set the `--authentication-kubeconfig` and `--authorization-kubeconfig` flags in the scheduler. These are already set in the controller-manager.

By setting them on the scheduler, we are telling it to perform tokenaccessreview and subjectaccessreview for any request coming in the secure port. This enables scraping /metrics on the secure port.

**Comment by joshrosso** [comment]
/remove-lifecycle stale

Still an important issue. I will work to make time to knock this out.

**Comment by luxas** [comment]
I think this checklist still holds pretty well: https://github.com/kubernetes/kubernetes/pull/72159#issuecomment-452393639

**Comment by neolit123** [comment]
@joshrosso PR needs rebase.

**Comment by neolit123** [comment]
we sent a PR for this here: https://github.com/kubernetes/kubernetes/pull/85043
/close

**Pull request closed without merging**